src/Figshare.jl

```diff
@@ -0,0 +1,74 @@
+struct Figshare <: DataRepo
+end
+
+base_url(::Figshare) = "https://api.figshare.com/v2/articles"
+
+function description(repo::Figshare, mainpage)
+    desc = mainpage["description"]
+    authors = [value["full_name"] for value in mainpage["authors"]]
+    author = format_authors(authors)
+    license = mainpage["license"]["name"] * " (" *mainpage["license"]["url"] * ")"
+    date = mainpage["published_date"]
+    paper = mainpage["citation"]
+    
+    escape_multiline_string("""
+    Author: $(author)
+    License: $(license)
+    Date: $(date)
+
+    $(desc)
+
+    Please cite this paper:
+    $(paper)
+    if you use this in your research.
+    """, "\$")
+end
+
+function get_urls(repo::Figshare, page)
+    urls = []
+    for i in page["files"]
+        push!(urls, i["download_url"])
+    end
+    urls
+end
+
+function get_checksums(repo::Figshare, page)
+    checksums = []
+    for i in page["files"]
+        push!(checksums, (:md5, i["computed_md5"]))
+    end
+    checksums
+end
+
+function data_fullname(::Figshare, mainpage)
+    mainpage["title"]
+end
+
+function website(repo::Figshare, mainpage_url, mainpage)
+    mainpage["url_public_html"]
+end
+
+function match_figshare(uri::String)
+    try
+        identifier = match(r"\d{7}", uri).match
+        return true, identifier
+    catch ErrorException
+        return false
+    end
+end
+
+function mainpage_url(repo::Figshare, dataname)
+    #We are making it work for both figshare id or doi
+    if match_doi(dataname)[1]
+        identifier = match_doi(dataname)[2]
+        down_url = base_url(repo) * "?doi=" * identifier
+        doi_page = JSON.parse(text_only(getpage(down_url).root))
+        url = doi_page[1]["url_public_api"]
+    elseif match_figshare(dataname)[1]
+        identifier = match_figshare(dataname)[2]
+        url = base_url(repo) * "/" *identifier   
+    else
+        error("Please use a valid url, DOI, or Figshare ID")
+    end
```

Comment:
How about leave `match_doi` returning just `nothing` or the DOI.
And handle it with dispatch (the julian way :-D)

Not tested but the following I think should work.
It define 2 local functions, each with 2 methods.
One for `::Void` and one for `::RegexMatch`

```
function mainpage_url(repo::Figshare, dataname)
	get_url_doi(::Void) = get_url_figshare(match_figshare(dataname))
	function get_url_doi(identifier_match::RegexMatch)
		identifier = identifier_match.match
        down_url = base_url(repo) * "?doi=" * identifier
        doi_page = JSON.parse(text_only(getpage(down_url).root))
		url = doi_page[1]["url_public_api"]
		return url
	end
	
	get_url_figshare(::Void) = error("Please use a valid url, DOI, or Figshare ID")
	function get_url_figshare(identifier_match::RegexMatch)
		identifier = identifier_match.match
		url = base_url(repo) * "/" *identifier
		return url
	end
	
        get_doi_url(match_doi(dataname))
end
```

You could also write that without using dispatch to local functions:

```
function mainpage_url(repo::Figshare, dataname)
	identifier_match = match_doi(dataname)
	if identifier_match isa RegexMatch
	    identifier = identifier_match.match
        down_url = base_url(repo) * "?doi=" * identifier
        doi_page = JSON.parse(text_only(getpage(down_url).root))
		url = doi_page[1]["url_public_api"]
		return url
	else
		identifier_match = match_figshare(dataname)
		if identifier_match isa RegexMatch
			identifier = identifier_match.match
			url = base_url(repo) * "/" *identifier
			return url
		else
			error("Please use a valid url, DOI, or Figshare ID")
		end
	end
end
```



Alternatively you could make those regular expressions just constants,
and then write ising `ismatch` and `match`.
In theory this is slower, but this is not performance critical,
and it tiny for a regex on such a short string.

```
function mainpage_url(repo::Figshare, dataname)
	if ismatch(doi_regex, dataname)
	    identifier = match(doi_regex, dataname).match
        down_url = base_url(repo) * "?doi=" * identifier
        doi_page = JSON.parse(text_only(getpage(down_url).root))
		url = doi_page[1]["url_public_api"]
		return url
	elseif ismatch(figshare_regex, dataname)
	    identifier = match(figshare_regex, dataname).match
		url = base_url(repo) * "/" *identifier
		return url
	else
		error("Please use a valid url, DOI, or Figshare ID")
	end
end
```


You could abstract this with a helper function, which might be useful elsewhere too
Though I am not sure that those big anon are args are very readabale.
I think this is over engineered.

```
function match_ifelse(pattern, str, matchedfunc, nomatchfunc)
	match_res = match(pattern, str)
	if match_res isa RegexMatch
		matchedfunc(match_res.match)
	else
		nomatchfunc()
	end
end


function mainpage_url(repo::Figshare, dataname)
	match_ifelse(doi_regex, dataname,
		function(identifier) # Matched DOI
			down_url = base_url(repo) * "?doi=" * identifier
			doi_page = JSON.parse(text_only(getpage(down_url).root))
			url = doi_page[1]["url_public_api"]
			return url
		end,
		function() #No match
			match_ifelse(figshare_regex, dataname,
				function(identifier) # Matched Figshare
					url = base_url(repo) * "/" *identifier
					return url
				end,
				function() # No match
					error("Please use a valid url, DOI, or Figshare ID")
				end
			)
		end
		)
	end
end
```



These are some options.
(I got a bit carried away).


Functions that sometimes return one argument and sometimes return  arguments,
are pretty unusual in julia,
and so are likely to catch future maintainers offguard.
The return type of `match` (`Union{RegexMatch, nothing}`) is a bit unusual, but less so and consistant with RegexMatch.